tests/unit/s2n_random_test.c

```diff
@@ -184,169 +111,338 @@ int main(int argc, char **argv)
+{
+    uint8_t data1[RANDOM_GENERATE_DATA_SIZE];
+    uint8_t data2[RANDOM_GENERATE_DATA_SIZE];
+    struct s2n_blob blob1 = { .data = data1 };
+    struct s2n_blob blob2 = { .data = data2 };
+
+    /* Get two sets of data in the same process/thread, and confirm that they
+     * differ
+     */
+    blob1.size = RANDOM_GENERATE_DATA_SIZE;
+    blob2.size = RANDOM_GENERATE_DATA_SIZE;
+    EXPECT_OK(s2n_get_public_random_data(&blob1));
+    EXPECT_OK(s2n_get_public_random_data(&blob2));
+    EXPECT_BYTEARRAY_NOT_EQUAL(data1, data2, RANDOM_GENERATE_DATA_SIZE);
+    EXPECT_OK(s2n_get_private_random_data(&blob1));
+    EXPECT_BYTEARRAY_NOT_EQUAL(data1, data2, RANDOM_GENERATE_DATA_SIZE);
+    EXPECT_OK(s2n_get_private_random_data(&blob2));
+    EXPECT_BYTEARRAY_NOT_EQUAL(data1, data2, RANDOM_GENERATE_DATA_SIZE);
+
+    return S2N_RESULT_OK;
+}
+
+static int s2n_common_tests(struct random_test_case *test_case)
+{
+    uint8_t data1[RANDOM_GENERATE_DATA_SIZE];
+    uint8_t data2[RANDOM_GENERATE_DATA_SIZE];
+    struct s2n_blob blob1 = { .data = data1 };
+    struct s2n_blob blob2 = { .data = data2 };
+
+    /* Get one byte of data, to make sure the pool is (almost) full */
+    blob1.size = 1;
+    blob2.size = 1;
+    EXPECT_OK(s2n_get_public_random_data(&blob1));
+    EXPECT_OK(s2n_get_private_random_data(&blob2));
+
+    /* Verify we generate unique data over threads */
+    EXPECT_OK(s2n_thread_test(s2n_get_public_random_data, GET_PUBLIC_RANDOM_DATA));
+    EXPECT_OK(s2n_thread_test(s2n_get_private_random_data, GET_PRIVATE_RANDOM_DATA));
+    EXPECT_OK(s2n_thread_test(s2n_get_public_random_data, GET_PRIVATE_RANDOM_DATA));
+    EXPECT_OK(s2n_thread_test(s2n_get_private_random_data, GET_PUBLIC_RANDOM_DATA));
+
+    /* Verify we generate unique data over forks */
+    EXPECT_OK(s2n_fork_test(s2n_get_private_random_data, GET_PRIVATE_RANDOM_DATA));
+    EXPECT_OK(s2n_fork_test(s2n_get_public_random_data, GET_PUBLIC_RANDOM_DATA));
+    EXPECT_OK(s2n_fork_test(s2n_get_public_random_data, GET_PRIVATE_RANDOM_DATA));
+    EXPECT_OK(s2n_fork_test(s2n_get_private_random_data, GET_PUBLIC_RANDOM_DATA));
+
+    /* Basic tests generating randomness */
+    EXPECT_OK(s2n_basic_generate_tests());
+
+    /* Verify that there are no trivially observable patterns in the output */
+    EXPECT_OK(s2n_basic_pattern_tests(s2n_get_public_random_data));
+    EXPECT_OK(s2n_basic_pattern_tests(s2n_get_private_random_data));
+
+    /* Just a sanity check and avoids cppcheck "unassignedVariable" errors.
+     * In future PRs this part will be expanded.
+     */
+    blob1.size = RANDOM_GENERATE_DATA_SIZE;
+    EXPECT_OK(s2n_get_public_random_data(&blob1));
+    blob2.size = RANDOM_GENERATE_DATA_SIZE;
+    EXPECT_OK(s2n_get_private_random_data(&blob2));
+    EXPECT_BYTEARRAY_NOT_EQUAL(data1, data2, RANDOM_GENERATE_DATA_SIZE);
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_random_test_case_default_cb(struct random_test_case *test_case)
+{
+    EXPECT_SUCCESS(s2n_init());
```

Comment:
If both test cases immediately call s2n_init, why do you need to use BEGIN_TEST_NO_INIT?

Reply:
Clarified offline: S2N_BEGIN_TEST_NO_INIT isn't necessary currently, but it will be once Torben starts enabling and disabling fork protection features.